File: pywbrgmm/numerical_Zk.py
import jax 
import jax.numpy as jnp
import numpy as np 
import scipy.stats as stats

from mcmc.sampler import rand_inv_gamma
from measure.wasserstein import wass_gauss


def numerical_Zk(K_max, dim, config, n_mc=100):
    ''' Numerical Computation of ZK as in algorithm 1
    '''
    g0 = config["g0"]
    a0 = config["a0"]
    b0 = config["b0"]
    tau = config["tau"]
    
    mu_mc = jnp.zeros((n_mc, K_max, dim))
    sig_mc = jnp.zeros((n_mc, K_max, dim, dim))

    mu_mc = stats.multivariate_normal.rvs(
        jnp.zeros(dim), tau**2 * jnp.eye(dim), (n_mc, K_max)
    ) 
    sig_mc = np.zeros((n_mc, K_max, dim, dim))
    for n in range(n_mc):
        for k in range(K_max):
            np.fill_diagonal(sig_mc[n, k], rand_inv_gamma(a0, b0, config))
    sig_mc = jnp.array(sig_mc) 

    gg = jnp.zeros((n_mc, K_max)) 
    for n in range(n_mc): 
        for k in range(K_max): 
            min_d = 1.
            for i in range(k):
                for j in range(i, k): 
                    d = wass_gauss(
                        mu_mc[n, i], sig_mc[n, i], mu_mc[n, j], sig_mc[n, j]
                    ) 
                    min_d = min(min_d, d/(g0+d))
            gg.at[n, k].set(min_d)  
     
    Z = jnp.log(jnp.mean(gg, axis=0))
    Z.at[0].set(0.)
    return Z

File: WassersteinBayesRepulsiveGMM/src/numerical_Zk.jl
# Numerical Computation of ZK as in algorithm 1
@inline function numerical_Zₖ(K_max, dim, config, k_prior; n_mc=1000)
    g₀ = config["g₀"]
    a₀ = config["a₀"]
    b₀ = config["b₀"]
    τ = config["τ"]
    
    μ_mc = zeros(dim, K_max, n_mc)
    Σ_mc = zeros(dim, dim, K_max, n_mc)
 
    @inbounds μ_mc[:, :, :] .= randn((dim, K_max, n_mc)) * τ
    
    Threads.@threads for n = 1:n_mc
        for k = 1:K_max
            Σ_mc[:, :, k, n] .= rand(k_prior.biw)
        end 
    end 

    gg = zeros(K_max, n_mc) 
    Threads.@threads for n = 1:n_mc
        for k = 2:K_max 
            min_d = 1.
            @inbounds for i = 1:(k - 1), j = (i + 1):k 
                d = wass_gauss(
                    μ_mc[:, i, n], Σ_mc[:, :, i, n], 
                    μ_mc[:, j, n], Σ_mc[:, :, j, n]) 
                min_d = min(min_d, d/(g₀+d))
            end
            gg[k, n] = min_d  
        end
    end 

    Z = log.(mean(gg; dims=2)) |> vec
    Z[1] = 0
    return Z
end